# INITIO DATA PROCESSING

## **BACKGROUND**

This script has been developed to perform some of the processes needed for the INITIO project, which aims to study the prevalence of resistance to inhibitor of Integrase among treatment-naive individuals with recent HIV infection in England. 

HIV samples are sequenced by Illumima platform after using a next generation (NGS) protocol of target-enrichment by sequence capture. The data is firstly processed using the bioiformatics pipeline [Genomancer](https://github.com/davidfbibby/genomancer_scripts) for HIV. The method generates the following outputs:

* **HIV_read_count.tsv**, summary of the input and output of the reads processed by Genomancer for each step of the pipeline.  

* **HIV_typonomers.zip**, compressed file with a collection of *CSV* files for each sequence generated containing the number of reads mathcing a specific HIV reference sequence and its frequency. The reads used for this process are the ones obtained after using several filters (trimmomatic, dehumanised reads, etc) and can offer an idea of the subtyping of the samples. However, usual tools for the subtyping of HIV are used to confirm the classification of the viruses. 

* **HIV_quasibams.zip**, compressed file tith *TABULAR* files generated by the PHE tool **QuasiBAM**. Each file consists of the results of the mapping of the reads to a particular reference sequence, giving information about the nucelotides at every single position with the frequency for the nucleotides and the one for the amino acids for each codon. This file is used to generate a consensus sequence at a particular frequency of nucleotides at each position. 

* **HIV_genomes.fas**, nucleotide sequences in *FASTA* format generated by QuasiBAM at a minimun base frequency of 15% to call a nucleotide in the consensus and a minimum depth of 30 reads for calling a position. 

* **HIV_post-run.zip**, compressed file with *NEX* files to check the phylogeny of the sequences generated by Genomancer. 

This process is performed in the High Performance Computing (HPC) cluster. The TABULAR files are re-analysed to generate FASTA files at frequency of 20 and 2 % using a depth of 100 reads in order to study majority and minority viral variants. If needed, it also creates an additional FASTA file at frequency of 20% but depth of 30 reads. This process is perform in the HPC cluster by this script after invoking QuasiBAM tool. The script also extracts information from the HIV_read_count.tsv, HIV_genomes.fas, TABULAR files and HIV_post-run.zip to produces a **Summary CSV report** with information about the number of contigs identified by genomancer, sequencing metrics (Median Depth, sequence length...) and whether the sequences are used to build the phylogeny. An extra **CSV report** is generated with the *Coverage Depth* for all the sequences that generated a TABULAR file by Genomancer. 

In a second phase, the query nucleotide sequences generated for each sample are initially aligned to a whole genome sequence of the reference sequence strain HXB2 (K03455.1). The resulting alignment is then **trimmed** to create *sub-alignments* for each gene using the coordinates of the reference sequence for each domain (Gag, Polymerase, Vif, Vpr, Tat1 and Tat2,Rev1 and Rev2, Vpu, Gp160 and Nef). The sequences for each *sub-alignment* and analysed for **potential frame-shifts** using the following algorithm. The trimmed amino acid reference sequence for each gene should be in frame after removing any gap and traslation to amino acids. Each query sequence is evaluated for three different frames. Each one of them are translated to amino acids after removing any gaps and aligned to the trimmed amino acid reference sequence again. The frame alignments are scored for similarities and only if the first one gets the higher score, the sequence is considered to be potentially in frame. In a subsequent step, an evaluation for the **presence of stop codons** along the query sequences is carried out by detection of the symbol "*" in any position of the sequence except for the last one. Information about potential frame shifts or stop codons in the sequences are recorded in a **TXT file**, which can be used to evaluate if the sequences needs to be manually modified. If needed, this process can be done using a sequence editor (i.e MEGA, Bioedit) and a **FASTA file** containing the query nucleotide sequences and trimmed ones corresponding to the genes of strain HXB2 in combination with a XLSX file. This file contains the same information of the TABULAR file but an extra column is added to refer the positions at the 20PC FASTA sequence. As the inclusion of insertions and deletions depends of the Depth choosen to create the consensus sequence, the positions in this new column are aproximate.          

In a next phase, once the quality of the sequences has been confirmed, specific **FASTA files** are created to perform the following analyes:

* **Antiviral resistance genotyping**, using the tool ["HIVdb Program: Sequence Analysis" from the Stanford University HIV Drug Resistance Database](https://hivdb.stanford.edu/hivdb/by-sequences/).

* **HIV Subtyping** using the tools:
    * [COntext-based Modeling for Expeditious Typing (**COMET**)](https://comet.lih.lu/)
    * [**REGA** HIV subtyping tool**](https://www.genomedetective.com/app/typingtool/hiv)
    * [Jumping profile Hidden Markov Model (**jpHMM**), locally installed](http://jphmm.gobics.de/download.html). 

In a last step, the information from these analyses in combination with the one obtained from the sequencing process are consolidated in a single XLSX file, for the user to access the results. 

## **REQUIREMENT**

This metodh uses mainly the following packages and Python libraries:
- Python 3.6.13
- Biopython 1.78
- Pandas 1.1.5
- Openpyxl 3.0.9
- mafft 7.310

but the complete list of software used for this method is summaried in the file called initio_environment.yml, see below. 

One of these distibutions ([Anaconda](https://www.anaconda.com/products/individual) or [Miniconda](https://docs.conda.io/en/latest/miniconda.html)) must be used for installing and managing most or all the software packages and other additional dependencies. [Git](https://git-scm.com/) is used to download the repository into the subdirectory **"Antiviral Unit/Project - INITiO"** from GitLab by typing in a terminal:

`git clone https://github.com/juanledesma78/initio_data_processing.git`

, which will create a new directory with the files:

- [**hiv_initio_project.py**](hiv_initio_project.py)
- [**initio_environment.yml**](initio_environment.yml),
- [**K03455.1_HXB2.fasta**](K03455.1_HXB2.fasta)

To install all the dependecies in an environement called **initio** using the *YML* file, enter in a terminal 

`
conda env create -f path/to/initio_environment.yml -p /home/phe.gov.uk/another.user/miniconda3/envs/initio
`

being the value after the argument -p the the *location where conda is installed* followed by *envs/initio*.

## **USAGE**

The script **hiv_initio_project.py** manages all the tools the method offers. The **conda environment initio** must be activated before using this method by typing in the terminal 

`conda activate initio`

The usage of the script is as followed: 

`python3 /parth/to/hiv_initio_project.py -i [-p] [-qc] [-c] [-f] [-d] [-m]`

with **-i** or **--input** taking the path to the run containing the data to analyse. This argument is ***REQUIRED*** for all the tools (selected by the specific optional arguments) to work. Details about the function of each tool can be found by typing:

`python3 hiv_initio_project.py --help`

and are explained in detail in the following secction:

* **-p, --postgenomancer**. This tool needs to be **run in the HPC cluster** to use the PHE QuasiBAM and to get access to the files *_quasibams.zip*,*_post-run.zip* and *_typonomer.zip*, *_genomes.fas* and *_read_counts.tsv* produced by Genomancer pipeline. It decompresses the files *"_quasibams.zip"*,*"_post-run.zip"* and *"_typonomer.zip"* into new subdirectories *quasibams*, *post-run* and *typonomer*, respectively. The files *.tabular* stored in *quasibams* are processed to generate a file *Coverage_Depth_Summary_HIV.csv*, which contains the coverage depth for all the sequences. Then QuasiBAM tool is invoked in the cluster to generate sequences of Majority and Minority variants (frequency of 20PC and 2PC) at depth of 100 reads.  The files *.nex* from directory *post-run*,  *_read_counts.tsv* and *_genomes.fas* are then used to create a file *Summary_Genomancer_Results_HIV.csv*, with a summary of the results obtained by Genomancer. Each fasta sequence is extracted from the file *_genomes.fas* into a new directory called *FASTAs*, in case the user needs it. All the files and folders labelled as "^" in the Figure 1 **must** be transferred then to the specific Run folder in the *Project - INITiO*. A value "*mad30*" can be used after *-p/--postgenomancer* to produce additional sequences at depth of 30 reads for Majority variants (20PC) for the same sample. 
                      
* **-qc, --quality**. Each *.tabular* file from the directory *quasibams* is used to generate a new file *_Quasibam_FASTA_numbering.xlsx*, used for potential inspection of positions in the sequences. The *FASTA files* from each sample located in the same directory are aligned to the reference sequence [**K03455.1_HXB2.fasta**](K03455.1_HXB2.fasta) and the resulting alignmets saved in a recently created subdirectory *quasibams/tmps*. These alignments are trimmed in sub-alignments corresponding to each genomic region and evaulated for frame shifts and/or stop codons. The information is then recorded in the file *FRAMESHIFT_initial_check_RUNID.txt* in *quasibams* (see example in Figure 2). In a subdirectory *quasibams/data_to_clean*, a file *_aligned_to_HXB2.fas* for each sample is created for potential sequence editing if needed, containing the query sequences aligned to the gene segments of reference.  

* **-c, --confirmation**. This tools works more or less similar to the one above but in this case just takes the *FASTA* files from the subdirectory *quasibams/data_to_clean* and repeats the same evaluation for frame-shifts/stop codons, generating a file *FRAMESHIFT_post_cleaning_RUNID.txt*.

* **-f, --files**. It takes the recently confirmed and clean *FASTA* files from the subdirectory *quasibams/data_to_clean* and creates the files *_2-20PC_D100_seqs_for_Resistance_report.fasta* and *_20PC_D100_seqs_for_WG_Subtyping.fasta* in a new directory called *fastas*, used for subtyping analysis and antiviral resistance genotyping. An additional file *_20PC_D30_seqs_for_All_Analyses.fasta* may be created if majoritiy variants at depth of 30 reads have been analysed. New directories *hivdb.stanford.report* and *subtyping* are generated to store the reports of the analyses to be done. Finally a file *_2-20PC_HIV_Genome_map.fasta* is created for numbering and genomic mapping purposes, containing the query sequences aligned to all the gene segments of reference.

* **-d, --data**. This tool decompressed the file *.zip* with the Stanford report from *hivdb.stanford.report* and consolidates the data with the information from *RUNID_sample_list.csv*, *RUNID_NGS_QC.csv*, *Summary_Genomancer_Results_HIV.csv*, and the *CSV files* from *subtyping/comet* and *subtyping/jphmm* for those sequences at depth of 100 reads and frequency of 20PC and 2PC (Majority and Minority Variants, respectively). It creates a file *_NGS_Results_2-20PC_D100.xlsx* in the main directory of the INITIO batch, with the information for all the runs analysed in the same batch. 

* **-m, --mayority**, it works similar to the argument "--data" but this is specific for those sequences generated at depth of 30 reads and frequency of 20PC (Majority).



<b>

**Figure 1**. Structure of the INITIO project directory. An initial batch (i.e. INITIO2019-2020) contains as many NGS runs as needed to process all the samples of the batch. The NGS run has several subdirectories created by any of the tools (#) and some, which are temporay, are removed when they are not needed any more (+). The symbol ^ indicates the folders and files than need to be transferred from the CLuster to the final destination. At the moment, the user needs to provide some files for some of the tools to carry out therir functions. The tools corresponding to the arguments are displayed on the right side pointing the element of the directory they work on.
</b>
```
.
└── INITIO_Project
    ├── Batch
    │   ├── INITIO_Batch_NGS_Results_2-20PC_D100.xlsx                       	<--- DATA
    │   └── NGS_Run
    │       ├── NGS_Run_NGS_QC.csv                         	<~~~~~~~~~~~~~~~~~~~ Provided by User
    │       ├── NGS_Run_sample_list.csv		                <~~~~~~~~~~~~~~~~~~~ Provided by User
    │       ├── Summary_Genomancer_Results_HIV.csv	(^)		        <--- POSTGENOMANCER
    │       ├── Coverage_Depth_Summary_HIV.csv (^)				<--- POSTGENOMANCER
    │       ├── typonomer (# ^)						        <--- POSTGENOMANCER
    │       │   └── 1000XX_Sample_ID.subtype.csv
    │       ├── HIV_genomes.fas
    │       ├── post-run (# ^)					                <--- POSTGENOMANCER
    │       │   ├── env.nex
    │       │   ├── gag.nex
    │       │   ├── nef.nex
    │       │   ├── pol.nex
    │       │   ├── vif.nex
    │       │   └── vpu.nex
    │       ├── quasibams (# ^)					                <--- POSTGENOMANCER
    │       │   ├── Sample_ID.1.20PC.fas				        <--- POSTGENOMANCER
    │       │   ├── Sample_ID.1.2PC.fas					        <--- POSTGENOMANCER
    │       │   ├── Sample_ID.1.tabular					        <--- POSTGENOMANCER
    │       │   ├── data_to_clean (+)					        <--- QUALITY
    │       │   │   └── Sample_ID.1_all_genes_aligned.fas                   	<--- QUALITY
    │       │   ├── FRAMESHIFT_initial_check_NGS_Run.txt                    	<--- QUALITY
    │       │   ├── Sample_ID.1_Quasibam_FASTA_numbering.xlsx               	<--- QUALITY
    │       │   ├── tmps (+)						        <--- QUALITY
    │       │   │   ├── Sample_ID.1_20-2PC_HXB2.fas			        <--- QUALITY
    │       │   │   └── Sample_ID.1_aligned_to_HXB2.fas			        <--- QUALITY 
    │       │   ├── FRAMESHIFT_post_cleaning_NGS_Run.txt		        <--- CONFIRMATION
    │       │   └── Sample_ID.1_2-20PC_HIV_Genome_map.fasta		        <--- FILES
    │       ├── hivdb.stanford.report (#)				        <--- FILES
    │       │   ├── NGS_Run.zi					<~~~~~~~~~~~~~~~~~~~ Provided by User
    │       │   └── NGS_Run							<--- DATA
    │       │       ├── MutationList_Sample_ID.1.20PC.csv
    │       │       ├── PrettyAA_IN.csv
    │       │       ├── PrettyAA_PR.csv
    │       │       ├── PrettyAA_RT.csv
    │       │       ├── Raw_Sample_ID.1.2PC.json
    │       │       ├── resistanceSummaries.csv
    │       │       ├── sequenceSummaries.csv
    │       │       └── unsequencedRegions.csv
    │       ├── fastas (#)						    	<--- FILES
    │       │   ├── NGS_Run_20PC_D100_seqs_for_WG_Subtyping.fasta	    	<--- FILES
    │       │   └── NGS_Run_2-20PC_D100_seqs_for_Resistance_report.fasta	<--- FILES
    │       ├── subtyping (#) 						    	<--- FILES
    │       │   ├── comet (#)
    │       │   │   └── results_NGS_Run.csv 		   	<~~~~~~~~~~~~~~~~~~~ Provided by User
    │       │   ├── jphmm (#)
    │       │   │   ├── Recombination_positions_NGS_Run_jpHMM.txt		<--- JPHMM
    │       │   │   └── results_NGS_Run_jpHMM.csv				<--- JPHMM
    │       │   ├── rega (#)
    │       │   │   └── results_NGS_Run.csv 		        <~~~~~~~~~~~~~~~~~~~ Provided by User
    │       │   └── rip
    │       ├── hivdb.stanford.report (#)				        <--- FILES
    │       │   ├── NGS_Run.zip				       	<~~~~~~~~~~~~~~~~~~~ Provided by User
    │       │   └── NGS_Run						        <--- DATA
    │       │       ├── MutationList_Sample_ID.1.20PC.csv
    │       │       ├── PrettyAA_IN.csv
    │       │       ├── PrettyAA_PR.csv
    │       │       ├── PrettyAA_RT.csv
    │       │       ├── Raw_Sample_ID.1.2PC.json
    │       │       ├── resistanceSummaries.csv
    │       │       ├── sequenceSummaries.csv
    │       │       └── unsequencedRegions.csv
    │       └── NGS_Run_Sequencing_results.csv					<--- DATA
    └── initio_data_processing
        ├── hiv_initio_project.py
        ├── initio_environment.yml
        └── K03455.1_HXB2.fasta
```



<b>

**Figure 2**. Example of report generated after using the tool invoke with argument -qc/--quality for checking potential frame shifts due to insertions or deletions and the presence of stop codons in the sequences. The tool was developed as an alternative to [HIV Sequence Locator](https://www.hiv.lanl.gov/content/sequence/LOCATE/locate.html)
![alt text](Frames_StopCodon_Report.png)
</b>




## **CONTACT DETAILS**

Juan Ledesma

UK Health Security Agency

juan.ledesma@ukhsa.gov.uk 

July 2022